# Goal of the analysis

- Check rates of various cols
- Fare comparison
- Survival rate
- Guess survival status in test data
    - age
    - gender
    - pclass
    - sex
    - fare

# Preprocessing

## Load data

In [1]:
# import library
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import seaborn as sns
sns.set_theme(style='whitegrid')

import sklearn
from sklearn.model_selection import KFold, train_test_split

In [2]:
# load data
submission = pd.read_csv('./data/gender_submission.csv')
test = pd.read_csv('./data/test.csv')
train = pd.read_csv('./data/train.csv')

## Train data

In [3]:
train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [4]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [5]:
train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

## Null values

In [6]:
# quick check of the null vals in train data
train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

##### Can see from above that train dataset has age: 177 null vals, cabin: 687 null vals, embarked: 2 null vals

### Cabin data

In [7]:
# too many null values in cabin col and no consistency of dataflow, delete.
train = train.drop(columns='Cabin', axis=1)

### Embarked data

In [8]:
# check null vals in Embarked col
# data indicates that both passengers are 1st class
train[train['Embarked'].isnull()]

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Embarked  
61   female  38.0      0      0  113572  80.0      NaN  
829  female  62.0      0      0  113572  80.0      NaN

In [9]:
# check the categorical amt of embarked, who are female 1st class passengers
train[(train['Pclass']==1)&(train['Sex']=='female')]['Embarked'].value_counts()

S    48
C    43
Q     1
Name: Embarked, dtype: int64

##### Embarked data from https://www.encyclopedia-titanica.org/titanic-first-class-passengers/

In [10]:
# replace nulls vals with 'S' in embarked col
train['Embarked'] = train['Embarked'].fillna('S')

### Age data

In [11]:
# check null vals
train[train['Age'].isnull()]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Embarked  
5      male  NaN      0      0      330877   8.4583        Q  
17     male  NaN      0      0      244373  13.0000        S  
19   female  NaN      0      0        2649   7.2250        C  
26     male  NaN      0      0        2631   7.2250        C  
28   female  NaN      0      0      330959   7.8792        Q  
..      ...  ...    ...    ...         ...      ...      ...  
859    male  NaN      0      0        2629   7.2292        C  
863  female  NaN      8      2    CA. 2343  69.5500        S  
868    male  NaN      0      0      345777   9.5000        S  
878    male  NaN      0      0      349217   7.8958        S  
888  female  NaN      1      2  W./C. 6607  23.4500        S  

[177 rows x 11 columns]

##### Age will be difference based on pclass

In [12]:
train[train['Pclass']==1].describe()

PassengerId    Survived  Pclass         Age       SibSp       Parch  \
count   216.000000  216.000000   216.0  186.000000  216.000000  216.000000   
mean    461.597222    0.629630     1.0   38.233441    0.416667    0.356481   
std     246.737616    0.484026     0.0   14.802856    0.611898    0.693997   
min       2.000000    0.000000     1.0    0.920000    0.000000    0.000000   
25%     270.750000    0.000000     1.0   27.000000    0.000000    0.000000   
50%     472.000000    1.000000     1.0   37.000000    0.000000    0.000000   
75%     670.500000    1.000000     1.0   49.000000    1.000000    0.000000   
max     890.000000    1.000000     1.0   80.000000    3.000000    4.000000   

             Fare  
count  216.000000  
mean    84.154687  
std     78.380373  
min      0.000000  
25%     30.923950  
50%     60.287500  
75%     93.500000  
max    512.329200

In [13]:
train[train['Pclass']==1]['Sex'].value_counts()

male      122
female     94
Name: Sex, dtype: int64

In [14]:
train[(train['Pclass']==1)&(train['Sex']=='male')]['Age'].mean()

41.28138613861386

In [15]:
train[(train['Pclass']==1)&(train['Sex']=='male')]['Age'].median()

40.0

In [16]:
train[(train['Pclass']==1)&(train['Sex']=='male')]['Age'].mode()

0    36.0
dtype: float64

In [17]:
train[(train['Pclass']==1)&(train['Sex']=='female')]['Age'].mean()

34.61176470588235

In [18]:
train[(train['Pclass']==1)&(train['Sex']=='female')]['Age'].median()

35.0

In [19]:
train[(train['Pclass']==1)&(train['Sex']=='female')]['Age'].mode()

0    35.0
dtype: float64

In [36]:
train.loc[(train['Pclass']==1)&(train['Sex']=='male')]['Age'].fillna(train[(train['Pclass']==1)&(train['Sex']=='male')]['Age'].median(), inplace=True)

/opt/conda/lib/python3.9/site-packages/pandas/core/generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


In [29]:
# df.loc[df['a'] == 1,'b'].fillna(df['b'].median(), inplace=True)
train.loc[(train['Pclass']==1)&(train['Sex']=='male')]['Age'].isnull() = train[(train['Pclass']==1)&(train['Sex']=='male')]['Age'].fillna(train[(train['Pclass']==1)&(train['Sex']=='male')]['Age'].median())

SyntaxError: cannot assign to function call (1677820910.py, line 2)

In [25]:
train[train['Age'].isnull()]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Embarked  
5      male  NaN      0      0      330877   8.4583        Q  
17     male  NaN      0      0      244373  13.0000        S  
19   female  NaN      0      0        2649   7.2250        C  
26     male  NaN      0      0        2631   7.2250        C  
28   female  NaN      0      0      330959   7.8792        Q  
..      ...  ...    ...    ...         ...      ...      ...  
859    male  NaN      0      0        2629   7.2292        C  
863  female  NaN      8      2    CA. 2343  69.5500        S  
868    male  NaN      0      0      345777   9.5000        S  
878    male  NaN      0      0      349217   7.8958        S  
888  female  NaN      1      2  W./C. 6607  23.4500        S  

[177 rows x 11 columns]

In [40]:
train['Age'] = train['Age'].fillna(train[(train['Pclass']==1)&(train['Sex']=='male')]['Age'].median())

In [41]:
train.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       2
dtype: int64

In [ ]:
train[train['Age']] = train['Age'].fillna()

In [16]:
train[train['Pclass']==2].describe()

PassengerId    Survived  Pclass         Age       SibSp       Parch  \
count   184.000000  184.000000   184.0  173.000000  184.000000  184.000000   
mean    445.956522    0.472826     2.0   29.877630    0.402174    0.380435   
std     250.852161    0.500623     0.0   14.001077    0.601633    0.690963   
min      10.000000    0.000000     2.0    0.670000    0.000000    0.000000   
25%     234.500000    0.000000     2.0   23.000000    0.000000    0.000000   
50%     435.500000    0.000000     2.0   29.000000    0.000000    0.000000   
75%     668.000000    1.000000     2.0   36.000000    1.000000    1.000000   
max     887.000000    1.000000     2.0   70.000000    3.000000    3.000000   

             Fare  
count  184.000000  
mean    20.662183  
std     13.417399  
min      0.000000  
25%     13.000000  
50%     14.250000  
75%     26.000000  
max     73.500000

In [17]:
train[train['Pclass']==3].describe()

PassengerId    Survived  Pclass         Age       SibSp       Parch  \
count   491.000000  491.000000   491.0  355.000000  491.000000  491.000000   
mean    439.154786    0.242363     3.0   25.140620    0.615071    0.393075   
std     264.441453    0.428949     0.0   12.495398    1.374883    0.888861   
min       1.000000    0.000000     3.0    0.420000    0.000000    0.000000   
25%     200.000000    0.000000     3.0   18.000000    0.000000    0.000000   
50%     432.000000    0.000000     3.0   24.000000    0.000000    0.000000   
75%     666.500000    0.000000     3.0   32.000000    1.000000    0.000000   
max     891.000000    1.000000     3.0   74.000000    8.000000    6.000000   

             Fare  
count  491.000000  
mean    13.675550  
std     11.778142  
min      0.000000  
25%      7.750000  
50%      8.050000  
75%     15.500000  
max     69.550000

SyntaxError: invalid syntax (1037885886.py, line 1)

In [12]:
train[train['Fare']==0]

PassengerId  Survived  Pclass                              Name   Sex  \
179          180         0       3               Leonard, Mr. Lionel  male   
263          264         0       1             Harrison, Mr. William  male   
271          272         1       3      Tornquist, Mr. William Henry  male   
277          278         0       2       Parkes, Mr. Francis "Frank"  male   
302          303         0       3   Johnson, Mr. William Cahoone Jr  male   
413          414         0       2    Cunningham, Mr. Alfred Fleming  male   
466          467         0       2             Campbell, Mr. William  male   
481          482         0       2  Frost, Mr. Anthony Wood "Archie"  male   
597          598         0       3               Johnson, Mr. Alfred  male   
633          634         0       1     Parr, Mr. William Henry Marsh  male   
674          675         0       2        Watson, Mr. Ennis Hastings  male   
732          733         0       2              Knight, Mr. Robert J  male   
806          807         0       1            Andrews, Mr. Thomas Jr  male   
815          816         0       1                  Fry, Mr. Richard  male   
822          823         0       1   Reuchlin, Jonkheer. John George  male   

      Age  SibSp  Parch  Ticket  Fare Embarked  
179  36.0      0      0    LINE   0.0        S  
263  40.0      0      0  112059   0.0        S  
271  25.0      0      0    LINE   0.0        S  
277   NaN      0      0  239853   0.0        S  
302  19.0      0      0    LINE   0.0        S  
413   NaN      0      0  239853   0.0        S  
466   NaN      0      0  239853   0.0        S  
481   NaN      0      0  239854   0.0        S  
597  49.0      0      0    LINE   0.0        S  
633   NaN      0      0  112052   0.0        S  
674   NaN      0      0  239856   0.0        S  
732   NaN      0      0  239855   0.0        S  
806  39.0      0      0  112050   0.0        S  
815   NaN      0      0  112058   0.0        S  
822  38.0      0      0   19972   0.0        S

### Test data

In [17]:
test

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

In [18]:
test.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200

In [19]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In [20]:
test.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64